File: app/src/main/java/com/example/changfeng/taptapword/MainActivity.java
package com.example.changfeng.taptapword;

import android.app.ActivityManager;


import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.design.widget.NavigationView;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import com.example.changfeng.taptapword.ui.ArchivedWordsFragment;
import com.example.changfeng.taptapword.ui.UnArchiveWordsFragment;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class MainActivity extends AppCompatActivity
        implements NavigationView.OnNavigationItemSelectedListener {


    private static final String TAG = "MainActivity";
    private static final String FRAGMENT_TAG = "CURRENT_FRAGMENT";

    private DrawerLayout drawerLayout;
    private ActionBarDrawerToggle drawerToggle;
    private boolean isWatchOn = false;

    @Bind(R.id.fab)
    FloatingActionButton addFab;

    @OnClick(R.id.fab)
    public void addWord() {
        Intent i = new Intent(getApplicationContext(), ConsultWordActivity.class);
        i.putExtra(ConsultWordActivity.TYPE, ConsultWordActivity.TYPE_CONSULT);
        startActivity(i);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.setDrawerListener(toggle);
        toggle.syncState();

        NavigationView navigationView = (NavigationView) findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);

        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);
        if (!prefs.getBoolean(SharedPref.FIRST_TIME, false)) {
            SharedPreferences.Editor editor = prefs.edit();
            editor.putBoolean(SharedPref.FIRST_TIME, true);

            editor.putBoolean(SharedPref.YOUDAO_DICT, true);
            editor.putBoolean(SharedPref.WEB_EXPLAIN, true);
            editor.putBoolean(SharedPref.YOUDAO_DICT, true);
            editor.putBoolean(SharedPref.BAIDU_TRANSLATE, true);

            editor.apply();

            setupFragment(new HelpFragment());

            Calendar c = Calendar.getInstance();
            int year = c.get(Calendar.YEAR);
            int month = c.get(Calendar.MONTH) + 1;
            int date = c.get(Calendar.DAY_OF_MONTH);
            int hour = c.get(Calendar.HOUR_OF_DAY);
            int minute = c.get(Calendar.MINUTE);
            int second = c.get(Calendar.SECOND);

            Word word = new Word();
            word.setName(getString(R.string.word_word_name));
            word.setAmPhone(getString(R.string.word_word_ph_am));
            word.setEnPhone(getString(R.string.word_word_ph_en));
            word.setMeans(getString(R.string.word_word_means));
            word.setArchived(false);
            word.setYear(year);
            word.setMonth(month);
            word.setDate(date);
            word.setHour(hour);
            word.setMinute(minute);
            word.setSecond(second);

            Word ninjaWord = new Word();
            ninjaWord.setName(getString(R.string.word_ninja_name));
            ninjaWord.setAmPhone(getString(R.string.word_ninja_ph_am));
            ninjaWord.setEnPhone(getString(R.string.word_ninja_ph_en));
            ninjaWord.setMeans(getString(R.string.word_ninja_means));
            ninjaWord.setArchived(false);
            ninjaWord.setYear(year);
            ninjaWord.setMonth(month);
            ninjaWord.setDate(date);
            ninjaWord.setHour(hour);
            ninjaWord.setMinute(minute);
            ninjaWord.setSecond(second);

            Word recentWord = new Word();
            recentWord.setName(getString(R.string.word_recent_name));
            recentWord.setAmPhone(getString(R.string.word_recent_ph_am));
            recentWord.setEnPhone(getString(R.string.word_recent_ph_en));
            recentWord.setMeans(getString(R.string.word_recent_means));
            recentWord.setArchived(false);
            recentWord.setYear(year);
            recentWord.setMonth(month);
            recentWord.setDate(date);
            recentWord.setHour(hour);
            recentWord.setMinute(minute);
            recentWord.setSecond(second);

            Word archiveWord = new Word();
            archiveWord.setName(getString(R.string.word_archive_name));
            archiveWord.setAmPhone(getString(R.string.word_archive_ph_am));
            archiveWord.setEnPhone(getString(R.string.word_archive_ph_en));
            archiveWord.setMeans(getString(R.string.word_archive_means));
            archiveWord.setArchived(false);
            archiveWord.setYear(year);
            archiveWord.setMonth(month);
            archiveWord.setDate(date);
            archiveWord.setHour(hour);
            archiveWord.setMinute(minute);
            archiveWord.setSecond(second);

            Word deleteWord = new Word();
            deleteWord.setName(getString(R.string.word_delete_name));
            deleteWord.setAmPhone(getString(R.string.word_delete_ph_am));
            deleteWord.setEnPhone(getString(R.string.word_delete_ph_en));
            deleteWord.setMeans(getString(R.string.word_delete_means));
            deleteWord.setArchived(false);
            deleteWord.setYear(year);
            deleteWord.setMonth(month);
            deleteWord.setDate(date);
            deleteWord.setHour(hour);
            deleteWord.setMinute(minute);
            deleteWord.setSecond(second);

            Word unarchiveWord = new Word();
            unarchiveWord.setName(getString(R.string.word_unarchive_name));
            unarchiveWord.setAmPhone(getString(R.string.word_unarchive_ph_am));
            unarchiveWord.setEnPhone(getString(R.string.word_unarchive_ph_en));
            unarchiveWord.setMeans(getString(R.string.word_unarchive_means));
            unarchiveWord.setArchived(true);
            unarchiveWord.setYear(year);
            unarchiveWord.setMonth(month);
            unarchiveWord.setDate(date);
            unarchiveWord.setHour(hour);
            unarchiveWord.setMinute(minute);
            unarchiveWord.setSecond(second);

            Word delete2Word = new Word();
            delete2Word.setName(getString(R.string.word_delete2_name));
            delete2Word.setAmPhone(getString(R.string.word_delete2_ph_am));
            delete2Word.setEnPhone(getString(R.string.word_delete2_ph_en));
            delete2Word.setMeans(getString(R.string.word_delete2_means));
            delete2Word.setArchived(true);
            delete2Word.setYear(year);
            delete2Word.setMonth(month);
            delete2Word.setDate(date);
            delete2Word.setHour(hour);
            delete2Word.setMinute(minute);
            delete2Word.setSecond(second);


            List<Word> words = new ArrayList<>();
            words.add(delete2Word);
            words.add(unarchiveWord);
            words.add(deleteWord);
            words.add(archiveWord);
            words.add(recentWord);
            words.add(ninjaWord);
            words.add(word);

            WordManger.get(getApplicationContext()).insertWords(words);


        }
        setupFragment(getLastFragmentId());

        startClipboardService();
        showToast("单词忍者正在监听");
        try {
            if (!prefs.getBoolean(SharedPref.INSTALL_SHORTCUT, false)) {
                addShortcut();
                prefs.edit().putBoolean(SharedPref.INSTALL_SHORTCUT, true).apply();
            }
        } catch (Exception e) {
            showToast(R.string.message_failed_install_shortcut);
        }

        countRunTimes();
        if (isToShowRateDialog()) {
            showRateDialog();
        }
    }


    @Override
    public void onBackPressed() {
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }


    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        // Handle navigation view item clicks here.
        int id = item.getItemId();
        setupFragment(id);
        return true;
    }

    private void setupFragment(int id) {
        if (id != R.id.recent_words) {
            addFab.setVisibility(View.GONE);
        }
        if (id == R.id.archive_words) {
            setupFragment(new ArchivedWordsFragment());
            setTitle(getString(R.string.menu_words));
            saveLastPageId(id);
//        } else if (id == R.id.achievement) {
//            Intent intent = new Intent(this, AchievementActivity.class);
//            startActivity(intent);
//            saveLastPageId(R.id.achievement);
        } else if (id == R.id.watch) {
            if (!isWatchOn) {
                startClipboardService();
                showToast("单词忍者正在监听");
            } else {
                stopClipboardService();
                showToast("单词忍者停止监听");
            }
            isWatchOn = !isWatchOn;
        } else if (id == R.id.settings)

        {
            setupFragment(new SettingsFragment());
            setTitle(getString(R.string.menu_settings));
            saveLastPageId(id);
        } else if (id == R.id.help)

        {
            setupFragment(new HelpFragment());
            setTitle(R.string.menu_help);
            saveLastPageId(id);
        } else if (id == R.id.rate) {
            rateUs();
        } else if (id == R.id.share) {
            shareByIntent();
        } else if (id == R.id.contact_us) {
            sendMailByIntent();
        } else if (id == R.id.about) {
            setupFragment(new AboutFragment());
            setTitle(R.string.title_about);
            saveLastPageId(id);
        } else {
            setupFragment(new UnArchiveWordsFragment());
            setTitle(getString(R.string.menu_recent_words));
            if (addFab.getVisibility() != View.VISIBLE) {
                addFab.setVisibility(View.VISIBLE);
            }
            saveLastPageId(R.id.recent_words);
        }

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
    }

    private void setupFragment(Fragment fragment) {
        FragmentManager fragmentManager = getSupportFragmentManager();
        Fragment currentFragment = fragmentManager.findFragmentByTag(FRAGMENT_TAG);

        if (currentFragment == null || !currentFragment.getClass().equals(fragment.getClass())) {
            fragmentManager.beginTransaction()
                    .replace(R.id.content_frame, fragment, FRAGMENT_TAG)
                    .commit();
        }
    }

    void startClipboardService() {
        stopClipboardService();
        Intent startIntent = new Intent(MainActivity.this, ClipboardService.class);
        startService(startIntent);
    }

    void stopClipboardService() {
        Intent stopIntent = new Intent(MainActivity.this, ClipboardService.class);
        stopService(stopIntent);
    }

    public static boolean isServiceRunning(Context mContext, String className) {
        boolean isRunning = false;
        ActivityManager activityManager = (ActivityManager)
                mContext.getSystemService(Context.ACTIVITY_SERVICE);
        List<ActivityManager.RunningServiceInfo> serviceList
                = activityManager.getRunningServices(30);
        if (!(serviceList.size() > 0)) {
            return false;
        }
        for (int i = 0; i < serviceList.size(); i++) {
            if (serviceList.get(i).service.getClassName().equals(className)) {
                isRunning = true;
                break;
            }
        }
        return isRunning;
    }

    private void openWebPage(String url) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        Uri uri = Uri.parse(url);
        intent.setData(uri);
        startActivity(intent);
    }

    private void rateUs() {
        try {
            Intent i = new Intent(Intent.ACTION_VIEW);
            i.setData(Uri.parse("market://details?id=" + getPackageName()));
            startActivity(i);
        } catch (Exception e) {
            showToast(getString(R.string.message_cannot_find_app_market));
        }
    }

    private void shareByIntent() {
        Intent intent = new Intent(Intent.ACTION_SEND);

        intent.setType("text/plain");
        intent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.share_subject));
        intent.putExtra(Intent.EXTRA_TEXT, getString(R.string.share_text) + " " + getString(R.string.uri_download_wandoujia));
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(Intent.createChooser(intent, "分享给你的亲友们"));
    }

    public void sendMailByIntent() {
        Uri uri = Uri.parse("mailto:" + getString(R.string.mail_address));
        Intent intent = new Intent(Intent.ACTION_SENDTO, uri);
        intent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.mail_subject));
        intent.putExtra(Intent.EXTRA_TEXT, getString(R.string.mail_text));

        PackageManager pm = getPackageManager();
        List<ResolveInfo> activities = pm.queryIntentActivities(intent, 0);
        if (activities.size() > 0) {
            startActivity(Intent.createChooser(intent, "请选择你的邮箱应用"));
        } else {
            showToast(getString(R.string.message_no_email_apps_found));
        }

    }

    private void addShortcut() {
        Intent shortcut = new Intent("com.android.launcher.action.INSTALL_SHORTCUT");
        shortcut.putExtra(Intent.EXTRA_SHORTCUT_NAME, getString(R.string.app_name));
        shortcut.putExtra("duplicate", false);

        Intent shortcutIntent = new Intent(Intent.ACTION_MAIN);
        shortcutIntent.setClassName(this, this.getClass().getName());
        shortcut.putExtra(Intent.EXTRA_SHORTCUT_INTENT, shortcutIntent);
        Intent.ShortcutIconResource iconRes = Intent.ShortcutIconResource.fromContext(this, R.drawable.ic_launcher512_512_round);
        shortcut.putExtra(Intent.EXTRA_SHORTCUT_ICON_RESOURCE, iconRes);
        sendBroadcast(shortcut);
    }

    private boolean hasRated() {
        SharedPreferences pref = getSharedPreferences(SharedPref.NAME, MODE_PRIVATE);
        return pref.getBoolean(SharedPref.RATED, false);
    }

    private void setRated() {
        SharedPreferences pref = getSharedPreferences(SharedPref.NAME, MODE_PRIVATE);
        pref.edit().putBoolean(SharedPref.RATED, true).apply();
    }

    private boolean isToShowRateDialog() {
        return !hasRated() && getRunTimes() > 5;
    }

    private void showRateDialog() {
        new AlertDialog.Builder(this)
                .setTitle(getString(R.string.title_rate))
                .setMessage(getString(R.string.message_rate))
                .setPositiveButton(getString(R.string.button_yes), new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        rateUs();
                        setRated();

                    }
                })
                .setNegativeButton(getString(R.string.button_no), new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        setRated();
                    }
                })
                .show();
    }

    private void countRunTimes() {
        SharedPreferences pref = getSharedPreferences(SharedPref.NAME, MODE_PRIVATE);
        int count = pref.getInt(SharedPref.RUN_TIMES, 0);
        count++;
        pref.edit().putInt(SharedPref.RUN_TIMES, count).apply();
    }

    private int getRunTimes() {
        SharedPreferences pref = getSharedPreferences(SharedPref.NAME, MODE_PRIVATE);
        return pref.getInt(SharedPref.RUN_TIMES, 0);
    }

    private void saveLastPageId(int id) {
        SharedPreferences.Editor editor = getSharedPreferences(SharedPref.NAME, MODE_PRIVATE).edit();
        editor.putInt(SharedPref.LAST_PAGE_ID, id).apply();
    }

    private int getLastFragmentId() {
        SharedPreferences pref = getSharedPreferences(SharedPref.NAME, MODE_PRIVATE);
        return pref.getInt(SharedPref.LAST_PAGE_ID, SharedPref.INVALID_FRAGMENT_ID);
    }

    private void showToast(int resourceId) {
        Toast.makeText(getApplicationContext(), getString(resourceId), Toast.LENGTH_SHORT).show();
    }

    private void showToast(String info) {
        Toast.makeText(getApplicationContext(), info, Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/com/example/changfeng/taptapword/MainActivity.kt
package com.example.changfeng.taptapword


import android.app.ActivityManager
import android.app.AlertDialog
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.design.widget.NavigationView
import android.support.v4.app.Fragment
import android.support.v4.app.NotificationCompat
import android.support.v4.view.GravityCompat
import android.support.v4.widget.DrawerLayout
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.MenuItem
import android.view.View
import com.example.changfeng.taptapword.ui.ArchivedWordsFragment
import com.example.changfeng.taptapword.ui.UnArchiveWordsFragment
import com.umeng.analytics.MobclickAgent
import org.jetbrains.anko.*
import java.util.*

class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {

    private var isWatchOn = false

    val addFab: FloatingActionButton
        get() = find(R.id.fab)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        try {
            val id = if (intent.getBooleanExtra(NINJA_NEW_WORD, false)) R.id.recent_words else lastFragmentId
            saveLastPageId(id)
        } catch(e: Exception) {

        }

        if (defaultSharedPreferences.getBoolean(SharedPref.NOTIFICATION_NINJA_WATCH, true)) {
            val ninjaServiceNotificationBuilder = NotificationCompat.Builder(this).setSmallIcon(R.drawable.image_ninja).setContentTitle(getString(R.string.app_name)).setContentText("点击关闭忍者监听")
            val ninjaIntent = PendingIntent.getBroadcast(this, RESULT_OK, Intent(NINJA_BROADCAST).putExtra(NINJA_OPEN_FLAG, true), PendingIntent.FLAG_CANCEL_CURRENT)

            ninjaServiceNotificationBuilder.setContentIntent(ninjaIntent)
            notificationManager.notify(2, ninjaServiceNotificationBuilder.build())
        }


        addFab.onClick {
            startActivity(intentFor<ConsultWordActivity>(ConsultWordActivity.TYPE to ConsultWordActivity.TYPE_CONSULT))
        }

        val toolbar: Toolbar = find(R.id.toolbar)
        val drawer: DrawerLayout = find(R.id.drawer_layout)

        setSupportActionBar(toolbar)
        val toggle = ActionBarDrawerToggle(
            this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        drawer.setDrawerListener(toggle)
        toggle.syncState()

        val navigationView = findViewById(R.id.nav_view) as NavigationView
        navigationView.setNavigationItemSelectedListener(this)


        if (!defaultSharedPreferences.getBoolean(SharedPref.FIRST_TIME, false)) {
            val editor = defaultSharedPreferences.edit()
            editor.putBoolean(SharedPref.FIRST_TIME, true)

            editor.putBoolean(SharedPref.YOUDAO_DICT, true)
            editor.putBoolean(SharedPref.WEB_EXPLAIN, true)
            editor.putBoolean(SharedPref.YOUDAO_DICT, true)
            editor.putBoolean(SharedPref.BAIDU_TRANSLATE, true)
            editor.apply()

            setupFragment(HelpFragment())

            val c = Calendar.getInstance()
            val year = c.get(Calendar.YEAR)
            val month = c.get(Calendar.MONTH) + 1
            val date = c.get(Calendar.DAY_OF_MONTH)
            val hour = c.get(Calendar.HOUR_OF_DAY)
            val minute = c.get(Calendar.MINUTE)
            val second = c.get(Calendar.SECOND)

            val word = Word()
            word.name = getString(R.string.word_word_name)
            word.amPhone = getString(R.string.word_word_ph_am)
            word.enPhone = getString(R.string.word_word_ph_en)
            word.means = getString(R.string.word_word_means)
            word.isArchived = false
            word.year = year
            word.month = month
            word.date = date
            word.hour = hour
            word.minute = minute
            word.second = second

            val ninjaWord = Word()
            ninjaWord.name = getString(R.string.word_ninja_name)
            ninjaWord.amPhone = getString(R.string.word_ninja_ph_am)
            ninjaWord.enPhone = getString(R.string.word_ninja_ph_en)
            ninjaWord.means = getString(R.string.word_ninja_means)
            ninjaWord.isArchived = false
            ninjaWord.year = year
            ninjaWord.month = month
            ninjaWord.date = date
            ninjaWord.hour = hour
            ninjaWord.minute = minute
            ninjaWord.second = second

            val recentWord = Word()
            recentWord.name = getString(R.string.word_recent_name)
            recentWord.amPhone = getString(R.string.word_recent_ph_am)
            recentWord.enPhone = getString(R.string.word_recent_ph_en)
            recentWord.means = getString(R.string.word_recent_means)
            recentWord.isArchived = false
            recentWord.year = year
            recentWord.month = month
            recentWord.date = date
            recentWord.hour = hour
            recentWord.minute = minute
            recentWord.second = second

            val archiveWord = Word()
            archiveWord.name = getString(R.string.word_archive_name)
            archiveWord.amPhone = getString(R.string.word_archive_ph_am)
            archiveWord.enPhone = getString(R.string.word_archive_ph_en)
            archiveWord.means = getString(R.string.word_archive_means)
            archiveWord.isArchived = false
            archiveWord.year = year
            archiveWord.month = month
            archiveWord.date = date
            archiveWord.hour = hour
            archiveWord.minute = minute
            archiveWord.second = second

            val deleteWord = Word()
            deleteWord.name = getString(R.string.word_delete_name)
            deleteWord.amPhone = getString(R.string.word_delete_ph_am)
            deleteWord.enPhone = getString(R.string.word_delete_ph_en)
            deleteWord.means = getString(R.string.word_delete_means)
            deleteWord.isArchived = false
            deleteWord.year = year
            deleteWord.month = month
            deleteWord.date = date
            deleteWord.hour = hour
            deleteWord.minute = minute
            deleteWord.second = second

            val unarchiveWord = Word()
            unarchiveWord.name = getString(R.string.word_unarchive_name)
            unarchiveWord.amPhone = getString(R.string.word_unarchive_ph_am)
            unarchiveWord.enPhone = getString(R.string.word_unarchive_ph_en)
            unarchiveWord.means = getString(R.string.word_unarchive_means)
            unarchiveWord.isArchived = true
            unarchiveWord.year = year
            unarchiveWord.month = month
            unarchiveWord.date = date
            unarchiveWord.hour = hour
            unarchiveWord.minute = minute
            unarchiveWord.second = second

            val delete2Word = Word()
            delete2Word.name = getString(R.string.word_delete2_name)
            delete2Word.amPhone = getString(R.string.word_delete2_ph_am)
            delete2Word.enPhone = getString(R.string.word_delete2_ph_en)
            delete2Word.means = getString(R.string.word_delete2_means)
            delete2Word.isArchived = true
            delete2Word.year = year
            delete2Word.month = month
            delete2Word.date = date
            delete2Word.hour = hour
            delete2Word.minute = minute
            delete2Word.second = second


            val words = ArrayList<Word>()
            words.add(delete2Word)
            words.add(unarchiveWord)
            words.add(deleteWord)
            words.add(archiveWord)
            words.add(recentWord)
            words.add(ninjaWord)
            words.add(word)

            WordManger.get(applicationContext).insertWords(words)


        }
        setupFragment(lastFragmentId)

        startClipboardService()
        toast("单词忍者正在监听")
        try {
            if (!defaultSharedPreferences.getBoolean(SharedPref.INSTALL_SHORTCUT, false)) {
                addShortcut()
                defaultSharedPreferences.edit().putBoolean(SharedPref.INSTALL_SHORTCUT, true).apply()
            }
        } catch (e: Exception) {
            toast(R.string.message_failed_install_shortcut)
        }

        countRunTimes()
        if (isToShowRateDialog) {
            showRateDialog()
        }
    }

    override fun onResume() {
        super.onResume()
        MobclickAgent.onPageStart("主界面")
        MobclickAgent.onResume(this)
    }

    override fun onPause() {
        super.onPause()
        MobclickAgent.onPageEnd("主界面")
        MobclickAgent.onPause(this)
    }

    override fun onBackPressed() {
        val drawer = findViewById(R.id.drawer_layout) as DrawerLayout
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }


    @SuppressWarnings("StatementWithEmptyBody")
    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // Handle navigation view item clicks here.
        val id = item.itemId
        setupFragment(id)
        return true
    }

    private fun setupFragment(id: Int) {
        addFab.visibility = if (id == R.id.recent_words) View.VISIBLE else View.GONE
        when (id) {
            R.id.recent_words -> {
                setupFragment(UnArchiveWordsFragment())
                title = getString(R.string.menu_recent_words)
                saveLastPageId(id)
            }
            R.id.archive_words -> {
                setupFragment(ArchivedWordsFragment())
                title = getString(R.string.menu_words)
                saveLastPageId(id)
            }
            R.id.watch -> {
                if (!isWatchOn) {
                    startClipboardService()
                    toast("单词忍者正在监听")
                } else {
                    stopClipboardService()
                    toast("单词忍者停止监听")
                }
                isWatchOn = !isWatchOn
            }
            R.id.settings -> {
                setupFragment(SettingsFragment())
                title = getString(R.string.menu_settings)
                saveLastPageId(id)
            }
            R.id.help -> {
                setupFragment(HelpFragment())
                title = getString(R.string.menu_help)
                saveLastPageId(id)
            }
            R.id.rate -> rateUs()
            R.id.share -> {
                shareByIntent()
            }
            R.id.contact_us -> sendMailByIntent()
            R.id.about -> {
                setupFragment(AboutFragment())
                title = getString(R.string.menu_about)
                saveLastPageId(id)
            }
            else -> {
                setupFragment(UnArchiveWordsFragment())
                title = getString(R.string.menu_recent_words)
                addFab.visibility = View.VISIBLE
                saveLastPageId(R.id.recent_words)
            }
        }


        val drawer = findViewById(R.id.drawer_layout) as DrawerLayout
        drawer.closeDrawer(GravityCompat.START)
    }

    private fun setupFragment(fragment: Fragment) {
        val fragmentManager = supportFragmentManager
        val currentFragment = fragmentManager.findFragmentByTag(FRAGMENT_TAG)

        if (currentFragment == null || currentFragment.javaClass != fragment.javaClass) {
            fragmentManager.beginTransaction().replace(R.id.content_frame, fragment, FRAGMENT_TAG).commit()
        }
    }

    internal fun startClipboardService() {
        stopClipboardService()
        val startIntent = Intent(this@MainActivity, ClipboardService::class.java)
        startService(startIntent)
    }

    internal fun stopClipboardService() {
        val stopIntent = Intent(this@MainActivity, ClipboardService::class.java)
        stopService(stopIntent)
    }

    private fun rateUs() {
        try {
            val i = Intent(Intent.ACTION_VIEW)
            i.data = Uri.parse("market://details?id=" + packageName)
            startActivity(i)
        } catch (e: Exception) {
            toast(getString(R.string.message_cannot_find_app_market))
        }

    }

    private fun shareByIntent() {
        share(getString(R.string.share_text, getString(R.string.uri_download_wandoujia)), getString(R.string.share_subject))
    }

    fun sendMailByIntent() {
        try {
            email(getString(R.string.mail_address), getString(R.string.mail_subject), getString(R.string.mail_text))
        } catch (e: Exception) {
            toast(getString(R.string.message_no_email_apps_found))
        }
    }

    private fun addShortcut() {
        val shortcut = Intent("com.android.launcher.action.INSTALL_SHORTCUT")
        shortcut.putExtra(Intent.EXTRA_SHORTCUT_NAME, getString(R.string.app_name))
        shortcut.putExtra("duplicate", false)

        val shortcutIntent = Intent(Intent.ACTION_MAIN)
        shortcutIntent.setClassName(this, this.javaClass.name)
        shortcut.putExtra(Intent.EXTRA_SHORTCUT_INTENT, shortcutIntent)
        val iconRes = Intent.ShortcutIconResource.fromContext(this, R.drawable.ic_launcher512_512_round)
        shortcut.putExtra(Intent.EXTRA_SHORTCUT_ICON_RESOURCE, iconRes)
        sendBroadcast(shortcut)
    }

    private fun hasRated(): Boolean {
        return defaultSharedPreferences.getBoolean(SharedPref.RATED, false)
    }

    private fun setRated() {
        defaultSharedPreferences.edit().putBoolean(SharedPref.RATED, true).apply()
    }

    private val isToShowRateDialog: Boolean
        get() = !hasRated() && runTimes > 5

    private fun showRateDialog() {
        AlertDialog.Builder(this).setTitle(getString(R.string.title_rate)).setMessage(getString(R.string.message_rate)).setPositiveButton(getString(R.string.button_yes)) { dialog, which ->
            rateUs()
            setRated()
        }.setNegativeButton(getString(R.string.button_no)) { dialog, which -> setRated() }.show()
    }

    private fun countRunTimes() {
        var count = defaultSharedPreferences.getInt(SharedPref.RUN_TIMES, 0)
        count++
        defaultSharedPreferences.edit().putInt(SharedPref.RUN_TIMES, count).apply()
    }

    private val runTimes: Int
        get() = defaultSharedPreferences.getInt(SharedPref.RUN_TIMES, 0)

    private fun saveLastPageId(id: Int) {
        defaultSharedPreferences.edit().putInt(SharedPref.LAST_PAGE_ID, id).apply()
    }

    private val lastFragmentId: Int
        get() = defaultSharedPreferences.getInt(SharedPref.LAST_PAGE_ID, SharedPref.INVALID_FRAGMENT_ID)

    companion object {
        private val TAG = "MainActivity"
        private val FRAGMENT_TAG = "CURRENT_FRAGMENT"

        val NINJA_BROADCAST = "com.example.changfeng.taptapword.ninja"
        val NINJA_OPEN_FLAG = "ninja_open_flag"
        val NINJA_NEW_WORD = "new_word"

        fun isServiceRunning(mContext: Context, className: String): Boolean {
            var isRunning = false
            val activityManager = mContext.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
            val service = activityManager.getRunningServices(30).find { s -> s.service.className == className }
            return service != null
        }
    }
}
